src/misc/JsonConfig.cpp

```diff
@@ -0,0 +1,173 @@
+#include "JsonConfig.h"
+#include <filesystem>
+#include <fstream>
+#include <spdlog/spdlog.h>
+
+namespace fs = std::filesystem;
+
+JsonConfig::JsonConfig(std::string path) : mPath(path) {
+    // check if path exists and that it's a file
+    if (!fs::exists(path) || !fs::is_regular_file(path)) {
+        mIsNewConfig = true;
+        mJson = json::object();
+        return;
+    }
+
+    try {
+        std::ifstream ifs(path);
+        mJson = json::parse(ifs);
+    } catch (json::parse_error& e) {
+        SPDLOG_ERROR("Failed to parse JSON file: {}", e.what());
+
+        // If failure to parse, we will create a new file
+        mIsNewConfig = true;
+        mJson = json::object();
+    }
+}
+
+json::json_pointer JsonConfig::DotKeyToPointer(std::string key) {
+    std::replace(key.begin(), key.end(), '.', '/');
+    key = "/" + key;
+
+    return json::json_pointer(key);
+}
+
+void JsonConfig::SetArbitraryType(std::string key, json::value_type value) {
+    auto ptr = DotKeyToPointer(key);
+    mJson[ptr] = value;
+}
+
+json::value_type JsonConfig::GetArbitraryType(std::string key) {
+    auto ptr = DotKeyToPointer(key);
+
+    try {
+        return mJson.at(ptr);
+    } catch (json::out_of_range& e) { return nullptr; }
+}
+
+// MARK: - Public API
+
+bool JsonConfig::IsNewConfig() {
+    return mIsNewConfig;
+}
+
+void JsonConfig::DeleteEntry(std::string key) {
+    // nlohmann::json doesn't have a way to delete a json pointer so
+    // we'll work around it by grabbing the parent object of the key
+    // we want to delete and deleting the key
+
+    auto currentKey = key;
+    auto parentKey = key;
+
+    try {
+        bool shouldStop = false;
+        while (!shouldStop) {
+            // grab parent of current key
+            parentKey = currentKey.substr(0, currentKey.find_last_of('.'));
+            auto parentPtr = DotKeyToPointer(parentKey);
+
+            // grab parent object at key and delete the current key
+            auto& parent = mJson.at(parentPtr);
+            auto keyToDelete = currentKey.substr(currentKey.find_last_of('.') + 1);
+
+            // check if parent object is an array and delete index
+            if (parent.is_array()) {
+                auto index = std::stoi(keyToDelete);
+                parent.erase(index);
+            } else {
+                parent.erase(keyToDelete);
+            }
+
+            // set variable for whether we should keep recursing parents
+            shouldStop = parent.empty() || parentKey == currentKey;
+            currentKey = parentKey;
```

Comment:
I think the `parentKey == currentKey` check should be moved to the start of the while loop as an early break/return.

As it is now, if you pass in a key like `xxx.yyy`, it will delete `yyy` from `xxx`, but if `xxx` is not empty, it will loop one more time and attempt to do a delete for `xxx` on `xxx`. which would be bad if there was technically a real scenario of `xxx.xxx` containing a value.